io/EptAddonWriter.cpp

```diff
@@ -0,0 +1,252 @@
+{
+    for (const auto& addon : m_addons)
+    {
+        log()->get(LogLevel::Debug) << "Writing addon dimension " <<
+            addon->name() << " to " << addon->ep().prefixedRoot() << std::endl;
+
+        writeOne(view, *addon);
+
+        log()->get(LogLevel::Debug) << "\tWritten" << std::endl;
+    }
+}
+
+void EptAddonWriter::writeOne(const PointViewPtr view, const Addon& addon) const
+{
+    std::vector<std::vector<char>> buffers;
+    buffers.reserve(m_hierarchy.size());
+
+    // Create an addon buffer for each node we're going to write.
+    for (const auto& p : m_hierarchy)
+    {
+        buffers.emplace_back(p.second * addon.size(), 0);
+    }
+
+    // Fill in our buffers with the data from the view.
+    PointRef pr(*view);
+    uint64_t nodeId(0);
+    uint64_t pointId(0);
+    for (uint64_t i(0); i < view->size(); ++i)
+    {
+        pr.setPointId(i);
+        nodeId = pr.getFieldAs<uint64_t>(m_nodeIdDim);
+
+        // Node IDs are 1-based to distinguish points that do not come from the
+        // EPT reader.
+        if (!nodeId) continue;
+
+        nodeId -= 1;
+        pointId = pr.getFieldAs<uint64_t>(m_pointIdDim);
+
+        auto& buffer(buffers.at(nodeId));
+        assert(pointId * addon.size() + addon.size() <= buffer.size());
+        char* dst = buffer.data() + pointId * addon.size();
+        pr.getField(dst, addon.id(), addon.type());
+    }
+
+    const arbiter::Endpoint& ep(addon.ep());
+    const arbiter::Endpoint dataEp(ep.getSubEndpoint("ept-data"));
+    const arbiter::Endpoint hierEp(ep.getSubEndpoint("ept-hierarchy"));
+
+    if (ep.isLocal())
+    {
+        arbiter::fs::mkdirp(dataEp.root());
+        arbiter::fs::mkdirp(hierEp.root());
+    }
+
+    // Write the binary dimension data for the addon.
+    nodeId = 0;
+    for (const auto& p : m_hierarchy)
+    {
+        const Key key(p.first);
+
+        m_pool->add([&dataEp, &buffers, key, nodeId]()
+        {
+            dataEp.put(key.toString() + ".bin", buffers.at(nodeId));
+        });
+
+        ++nodeId;
+    }
+
+    m_pool->await();
+
+    // Write the addon hierarchy data.
+    Json::Value h;
+    const Key key(m_info->bounds());
+    writeHierarchy(h, Key(m_info->bounds()), hierEp);
+    hierEp.put(key.toString() + ".json", h.toStyledString());
+
+    m_pool->await();
+
+    // Write the top-level addon metadata.
+    Json::Value meta;
+    meta["type"] = getTypeString(addon.type());
+    meta["size"] = static_cast<Json::UInt64>(addon.size());
+    meta["version"] = "1.0.0";
+    meta["dataType"] = "binary";
+
+    ep.put("ept-addon.json", meta.toStyledString());
+}
+
+void EptAddonWriter::writeHierarchy(Json::Value& curr, const Key& key,
+        const arbiter::Endpoint& hierEp) const
+{
+    const std::string keyName(key.toString());
+    if (!m_hierarchy.count(keyName)) return;
+
+    const uint64_t np(m_hierarchy.at(keyName));
+    if (!np) return;
+
+    if (m_hierarchyStep && key.d && key.d == m_hierarchyStep)
```

Comment:
Seems like one of these &&'s is not needed.  Maybe the compiler is smart enough to see that.

Reply:
Fixed at 3736533.